### PR TITLE
[SPARK-44509][CONNECT][FOLLOW-UP] Add inheritable support in new job cancellation API

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR is a kind of a followup of https://github.com/apache/spark/pull/42120. This addresses my own comment at https://github.com/apache/spark/pull/42120#discussion_r1272933655 by adding the support of `InheritableThread` with Spark Connect.

### Why are the changes needed?

In order to allow the thread local hierarchy supported in Scala side. 

### Does this PR introduce _any_ user-facing change?

The main PR is not released out yet so it doesn't have user-facing change.

### How was this patch tested?

Unittests were added.

### CODE REVIEW

File: python/pyspark/util.py [hunk 1]
```diff
@@ -348,56 +373,87 @@ def wrapped(*args: Any, **kwargs: Any) -> Any:
 
 class InheritableThread(threading.Thread):
     """
-    Thread that is recommended to be used in PySpark instead of :class:`threading.Thread`
-    when the pinned thread mode is enabled. The usage of this class is exactly same as
-    :class:`threading.Thread` but correctly inherits the inheritable properties specific
-    to JVM thread such as ``InheritableThreadLocal``.
-
-    Also, note that pinned thread mode does not close the connection from Python
```
Comment (HyukjinKwon) [inline python/pyspark/util.py:356]: This is already fixed in Py4J, https://github.com/py4j/py4j/pull/471

**Comment by HyukjinKwon** [comment]
cc @zhengruifeng and @WeichenXu123 I need your look here.

**Review by zhengruifeng — APPROVED** [review]

**Review by WeichenXu123 — APPROVED** [review]
LGTM!

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs)

**Comment by HyukjinKwon** [comment]
All tests passed. The failed test is pip packaging test that happens after running all unittests. I will fix them separately at https://github.com/apache/spark/pull/42159.

Merged to master and branch-3.5.

**Pull request closed without merging**